zilliqa/src/crypto.rs

```diff
@@ -38,6 +38,25 @@ impl NodeSignature {
         Ok(NodeSignature(bls_signatures::aggregate(&signatures)?))
     }
 
+    // Verify that the aggregated signature is valid for the given public keys and message.
+    // That is, each public key has signed the message, and the aggregated signature is the
+    // aggregation of those signatures.
+    pub fn verify_aggregate(
+        signature: &NodeSignature,
+        message: &[u8],
+        public_keys: Vec<NodePublicKey>,
+    ) -> Result<()> {
+        // first verify that the signature itself has been composed from the public keys and the message
```

Comment:
Can we avoid forking `bls_signatures` by constructing the public key ourselves here?
e.g. something like
```rust
let aggregate_key: bls_signatures::PublicKey = public_keys.iter().map(|p| G1Projecttive::from(p.0)).sum().into();
aggregate_key.verify(signature.0, message);
```

Creating and maintaining a fork seems like quite a bit of effort, when I think we could just get away with implementing what we want directly here.

Reply:
No, there is a guard inside their library against verifying with multiple public keys signing the same message (preventing what we want to do), as it is a security vulnerability if you do it incorrectly.